facebook/react-native
# Fix portability issues to Linux, FreeBSD, and older libc++

Test Plan:
----------
I have a local Linux build (under Windows 10 WSL running Ubuntu 18.04) where I load an app bundle and ensure it doesn't crash, and run some of the gtest tests. I am currently unable to reproduce the cxxreact build in OSS, but this seems to be a known issue. For the record, my commandline for building and running the tests (after using vcpkg to install double-conversion, glog, gflags, and folly):
```clang++-6.0 -fuse-ld=gold -o cxxreact-test_clang6_O3_lto_marchnative_g3 -std=c++14 -O3 -flto -g3 -march=native -L/mnt/c/Users/matt/workspace/vcpkg/installed/x64-linux/lib -I/mnt/c/Users/matt/workspace/vcpkg/installed/x64-linux/include -I/usr/include/webkitgtk-4.0/ -I. -ljavascriptcoregtk-4.0 -lfolly -lglog -ldouble-conversion fabric/events/*.cpp fabric/events/tests/EventsTest.cpp fabric/debug/*.cpp fabric/debug/tests/DebugStringConvertibleTest.cpp cxxreact/*.cpp cxxreact/tests/*.cpp microprofiler/*.cpp  jsinspector/*.cpp  privatedata/*.cpp  jschelpers/JSC*.cpp jschelpers/[UV]*.cpp fabric/core/primitives/*.cpp fabric/core/tests/PrimitivesTest.cpp  -lgtest -lgtest_main /mnt/c/Users/matt/workspace/vcpkg/installed/x64-linux/lib/libfolly.a /mnt/c/Users/matt/workspace/vcpkg/installed/x64-linux/lib/libgtest*.a /mnt/c/Users/matt/workspace/vcpkg/installed/x64-linux/lib/libglog.a /mnt/c/Users/matt/workspace/vcpkg/installed/x64-linux/lib/libdouble-conversion.a /mnt/c/Users/matt/workspace/vcpkg/installed/x64-linux/lib/libgflags.a -lpthread```

Only two tests fail, which I will investigate and try to fix in a separate PR:
```
[----------] 2 tests from JSBigFileString
[ RUN      ] JSBigFileString.MapWholeFileTest
unknown file: Failure
C++ exception with description "basic_string::_M_construct null not valid" thrown in the test body.
[  FAILED  ] JSBigFileString.MapWholeFileTest (4 ms)
[ RUN      ] JSBigFileString.MapPartTest
unknown file: Failure
C++ exception with description "basic_string::_M_construct null not valid" thrown in the test body.
[  FAILED  ] JSBigFileString.MapPartTest (3 ms)
[----------] 2 tests from JSBigFileString (20 ms total)
```

Changelog:
--------------

[INTERNAL] [ENHANCEMENT] [ReactCommon/cxxreact] -Fix portability issues on Linux, FreeBSD, and older libc++ to enable fast unit and integration testing
* Fix missing includes that happen to work on some libc++ implementations due to transitive inclusion, like unistd.h
* Don't register hooks that rely on JSGlobalContextEnableDebugger and other things not available on all platforms (e.g. libjavascriptcoregtk-4.0) by guarding with __APPLE__ and __ANDROID__
* Add nullptr checks for optionally set global functions before using them
* add getGlobalVariable to enable finer-grained unit testing
